Print back trace from ERROR and ASSERT macros, add ERROR_NO_TRACE macro

nilsdeppe opened the pull request:
## Proposed changes

- Prints 10 frames (excluding the internals of the macros) when an ERROR or ASSERT is triggered
- This would be annoying for input file adjustments, so an `ERROR_NO_TRACE` macro is added and used in `PARSE_ERROR`
- Replaces the `sys::abort` call with `ERROR` in `std::set_terminate`. The most common place where a back trace is useful is when an uncaught exception causes `std::terminate` to be called.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Print backtrace on ERROR and ASSERT

- Prints a 10-frame backtrace when an ERROR or ASSERT is triggrred.
  We can adjust the name of frames if we need to, but I'm hopeful this
  gives sufficient context without being too overwhelming.
- The backtrace is printed _before_ the error message so that the error message
  is always the last thing in the output.
- A new macro, ERROR_NO_TRACE, is added and used in PARSE_ERROR. This does not
  print a backtrace. I am of the opinion that backtraces for input file errors
  are not very useful or user-friendly.
- There are likely other ERROR calls that should be converted to ERROR_NO_TRACE,
  but this PARSE_ERROR is the most obvious one.

nilsdeppe committed:
Switch set_terminate to call ERROR macro

The most common occurrence of std::terminate being called is uncaught
exceptions, which are exceptionally annoying to debug. Having a backtrace is
pretty much essential to have any idea where the uncaught exception came from.

wthrowe commented on src/Utilities/ErrorHandling/AbortWithErrorMessage.cpp:
```diff
@@ -4,24 +4,54 @@
+  constexpr size_t max_stack_depth_printed = 13;
+  // clang-tidy: We're using a C-interface so clang-tidy is very upset.
+  // NOLINTNEXTLINE
+  void* trace_elems[max_stack_depth_printed];
```
Use a `std::array` and pass `.data()` below.

wthrowe reviewed (changes requested):
I like the design decisions of printing the backtrace first and avoiding it entirely during option parsing.

nilsdeppe commented:
Thanks for the review @wthrowe!

I rebased on develop, changed `std::array` and switched the `char**` to a `std::unique_ptr<char*, free*>` to have that managed for us. I made these changes in a fixup

wthrowe commented:
Looks good.

nilsdeppe commented:
Okay, squashed. Thanks for the review!

moxcodes commented on src/Utilities/ErrorHandling/AbortWithErrorMessage.cpp:
```diff
@@ -3,25 +3,51 @@
+  int trace_elem_count = backtrace(trace_elems.data(), max_stack_depth_printed);
+  std::unique_ptr<char*, decltype(free)*> stack_syms{
+      backtrace_symbols(trace_elems.data(), trace_elem_count), free};
+  // Start at 2 to ignore stream operator and abort_with_error_message
```
comment says 2 but loop starts at 3

moxcodes commented on src/Utilities/ErrorHandling/AbortWithErrorMessage.hpp:
```diff
@@ -9,14 +9,21 @@
 
 /// \ingroup ErrorHandlingGroup
-/// Compose an error message and abort the program.
+/// Compose an error message including a back trace and abort the program.
```
some dox strings call it a `back trace`, others say `backtrace` - best to be consistent.

moxcodes reviewed (dismissed):
I only have extremely minor nitpicks. Please squash any changes immediately.
Thanks for adding this! I've verified that it appropriately gives a useful backtrace (some symbols end up undemangled, but that's a harder problem...)

nilsdeppe commented:
Updated and squashed. Thanks for the reviews :)

kidder commented:
@wthrowe please look at
